data/ai/micro_ais/micro_ai_helper.lua

```diff
@@ -262,7 +299,8 @@ function micro_ai_helper.micro_ai_setup(cfg, CA_parms, required_keys, optional_k
 
     -- Check whether there are invalid keys in the [micro_ai] tag
     for k in pairs(cfg) do
-        if (k ~= 'side') and (k ~= 'ai_type') and (k ~= 'action')
+        if (not showed_deprecation_message) -- otherwise this produces false positives
```

Comment:
I suspect this check isn't what you want it to be. It's true if there has _ever_ been a deprecation message raised for _any_ Micro AI, not whether there has been a deprecation message for the specific Micro AI being considered.

I'm not quite sure what the best approach is here. You could just skip numeric keys (no wait, you already are? why doesn't that work?), or you could change the deprecation message to be once per Micro AI (so `showed_deprecation_message` becomes `showed_deprecation_message[ai_type]`).

Reply:
> I'm not quite sure what the best approach is here. You could just skip numeric keys (no wait, you already are? why doesn't that work?), 

Because this is the loop over the [micro_ai] tag config, not over required_key or optional_keys.

> or you could change the deprecation message to be once per Micro AI (so `showed_deprecation_message` becomes `showed_deprecation_message[ai_type]`).

Hmm.  I actually figured that getting a deprecation message only once overall was good enough.  It gets reset at each reload, after all, and deprecation messages are for UMC authors while they work on their add-ons, not something you need to worry about while you use an add-on.

However, there's a related (bigger?) issue with this, in that if one UMC creator does not bother dealing with deprecation messages, that might disable these warnings for another author.  So maybe making it MAI type specific is the way to go.

Reply:
Ohh, it shows false positives because the code within this loop doesn't handle the old required/optional keys format? Is that the reasoning?

If that's the case, it feels like it'd be better to put the "if not showed_deprecation_message" outside the loop… though that adds another level of indentation…

Reply:
> Ohh, it shows false positives because the code within this loop doesn't handle the old required/optional keys format? Is that the reasoning?

Yes.

> If that's the case, it feels like it'd be better to put the "if not showed_deprecation_message" outside the loop… though that adds another level of indentation…

Yeah, I know.  I did it this way because if I put it around the outside, then that whole chunk of code has to have its indentation changed whenever this is removed in the future, even though it is just two little pieces of conditional that need to be removed and it will look like a much bigger change than it is.  And it's not like this is a performance issue.  But it is a little "less clean" this way, so I can move it outside if you prefer.

Reply:
Ah, you have a good point there. Fair enough, we can leave it like this.